### ISSUE TITLE
LR Scheduler bug when resuming training from checkpoint

### ISSUE THREAD

**Issue by clemsgrs** [pytorch/pytorch#60265]
## 🐛 Bug

When trying to resume training from a given checkpoint using the CosineAnnealingLR scheduler, there's a discontinuity in learning rate values that I can't explain nor fix!

Let’s say I trained a model for 45 epochs, then stopped training and saved both the optimizer state and the scheduler state.
I want to resume training from epoch 46, until epoch 100. To do that, I simply need to instantiate a new LR scheduler with last_epoch = 45. But when looking at learning rates values as a function of epochs, I get a discontinuity at epoch 45.

## To Reproduce

```
model = torch.nn.Sequential(
            OrderedDict(
                [
                    ('conv1', torch.nn.Conv2d(3, 20, 3)),
                    ('relu1', torch.nn.ReLU()),
                    ('conv2', torch.nn.Conv2d(20, 64, 3)),
                    ('relu2', torch.nn.ReLU()),
                ]
            )
        )

base_lr = 0.001

epoch_start = 0
epoch_stop = 45
epoch_end = 100

optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epoch_end, last_epoch=epoch_start-1)

# run 45 epochs and save learning rate values
lr_values = []
for i in range(epoch_start, epoch_stop):
    lr = scheduler.get_last_lr()[0]
    lr_values.append(lr)
    optimizer.step()
    scheduler.step()

# stop training and save both optimizer and scheduler states
optimizer_state = optimizer.state_dict()
scheduler_state = scheduler.state_dict()

# resume training 
epoch_restart = scheduler_state['last_epoch']
new_optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
new_optimizer.load_state_dict(optimizer_state)
new_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(new_optimizer, T_max=epoch_end, last_epoch=epoch_restart-1)

# run remaining epochs = []
for i in range(epoch_restart, epoch_end):
    lr = new_scheduler.get_last_lr()[-1]
    lr_values.append(lr)
    new_optimizer.step()
    new_scheduler.step()

# plot learning rate values as a function of epochs
plt.plot(range(epoch_start, epoch_end), lr_values)
plt.xlabel('epoch')
plt.ylabel('lr')
plt.title('resume training from checkpoint')
plt.show()
```

<a href="https://imgbb.com/"><img src="https://i.ibb.co/HXMrVDr/resume-from-checkpoint.png" alt="resume-from-checkpoint" border="0"></a>

You can notice a discontinuity at epoch 45

## Expected behavior

Let's run uninterrupted 100 epochs long training to see what would be the expected learning values

```
model = torch.nn.Sequential(
            OrderedDict(
                [
                    ('conv1', torch.nn.Conv2d(3, 20, 3)),
                    ('relu1', torch.nn.ReLU()),
                    ('conv2', torch.nn.Conv2d(20, 64, 3)),
                    ('relu2', torch.nn.ReLU()),
                ]
            )
        )

base_lr = 0.001

epoch_start = 0
epoch_end = 100

optimizer = torch.optim.Adam(model.parameters(), lr=base_lr)
scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(optimizer, T_max=epoch_end, last_epoch=epoch_start-1)

# run 100 epochs and save learning rate values
lr_values = []
for i in range(epoch_start, epoch_end):
    lr = scheduler.get_last_lr()[0]
    lr_values.append(lr)
    optimizer.step()
    scheduler.step()

# plot learning rate values as a function of epochs
plt.plot(range(epoch_start, epoch_end), lr_values)
plt.xlabel('epoch')
plt.ylabel('lr')
plt.title('one way training')
plt.show()
```

<a href="https://imgbb.com/"><img src="https://i.ibb.co/2qrnDnC/one-way-training.png" alt="one-way-training" border="0"></a>

## Environment

 - PyTorch Version: 1.5.1
 - Python version: Python 3.6.9

cc @vincentqb @jbschlosser @albanD @iramazanli

**Comment by emcastillo** [comment]
Hi,

I tried to reproduce this and effectively I see a discontinuity,
These are the lr values for range [45, 50) of the lr in the correct case:

```
[0.0005782172325201156, 0.0005480017749828651, 0.0005327962241219932, 0.0005175454467760945, 0.0005022644936211605]
```

And when running your script, the values where

```
[0.0005631470933157985, 0.0005480017749828651, 0.0005327962241219932, 0.000517545446
7760945, 0.0005022644936211605]
```

As you said there is a gap, the funny stuff is that the gap is gone if we also load the state dict of the scheduler (you don't do this in your example). But Since we are not doing anything with the scheduler in the middle, its state should be exactly the same but the initial calculated `lr` seems a bit off. Let me do some more research into it, as I agree this is a bug.

```
# Original scheduler state after the last step
{'T_max': 100, 'eta_min': 0, 'base_lrs': [0.001], 'last_epoch': 45, '_step_count': 46, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.0005782172325201156]}
# New scheduler
{'T_max': 100, 'eta_min': 0, 'base_lrs': [0.001], 'last_epoch': 45, '_step_count': 1, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.0005631470933157985]}
```

**Comment by emcastillo** [comment]
I found the issue behind this problem:

Let's say that the first loop (where we obtain the state), changes the lrs assigned to the parameters at the end of epoch 44.
And leaves it ready for epoch 45. (the scheduler step runs after the optimizer step).

Now we create a new scheduler and since the `last_epoch` is not -1, it will try to calculate the current lr by looking and the lr stored in the optimizers parameters groups. However, the optimizer parameter groups have the lr already updated for this epoch, so it's taking a `lr(step+1)` value as base lr when it should be `lr(step)`

The problem here is this `group['lr']` value, that is the current optimizer parameter lr
https://github.com/pytorch/pytorch/blob/master/torch/optim/lr_scheduler.py#L514

We can't rely on the lr stored in the optimizer, when creating a new scheduler, so what we should do is change the way in which the first lr for the scheduler is initialized.

I will send a PR

**Comment by clementgr** [comment]
Hi,

good catch! Thank you for the quick investigation.
Something else I found strange was: when initializing a new scheduler with `last_epoch != -1`, the optimizer's lr value gets modified. I'm not sure if this is an issue or expected behavior, and if this could be related to what you've just pointed out. 

Simpler reproducer:

```python
import torch

parameter = torch.nn.Parameter(torch.zeros(10), requires_grad=True)

base_lr = 0.001
epoch_start = 0
epoch_end = 100

optimizer = torch.optim.Adam([parameter], lr=base_lr)
print('1', optimizer.param_groups[0]['lr'])

scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    optimizer, T_max=epoch_end
)
print('2', scheduler.state_dict())

optimizer.step()
print('3', optimizer.param_groups[0]['lr'])
scheduler.step()
print('4', optimizer.param_groups[0]['lr'])
print('5', scheduler.state_dict())

optimizer_state = optimizer.state_dict()
new_optimizer = torch.optim.Adam([parameter], lr=base_lr)
new_optimizer.load_state_dict(optimizer_state)
print('6', new_optimizer.param_groups[0]['lr'])

new_scheduler = torch.optim.lr_scheduler.CosineAnnealingLR(
    new_optimizer, T_max=epoch_end, last_epoch=0
)
print('7', new_optimizer.param_groups[0]['lr'])
```

And this is the output:

```
1 0.001
2 {'T_max': 100, 'eta_min': 0, 'base_lrs': [0.001], 'last_epoch': 0, '_step_count': 1, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.001]}
3 0.001
4 0.0009997532801828658
5 {'T_max': 100, 'eta_min': 0, 'base_lrs': [0.001], 'last_epoch': 1, '_step_count': 2, 'verbose': False, '_get_lr_called_within_step': False, '_last_lr': [0.0009997532801828658]}
6 0.0009997532801828658
7 0.0009995066212363998
```

Note how the lr for the new optimizer goes from (6) to (7).
